### ISSUE TITLE
floor and ceil produce negative numbers for large inputs

### ISSUE THREAD

**Issue by wosc** [hashicorp/terraform#29100]
### Terraform Version
```
Terraform v1.0.1
```

### Terraform Configuration Files
None

### Expected Behavior
Rounding large numbers should not overflow(?) into negative numbers, since as far as I can tell from the [docs](https://www.terraform.io/docs/language/expressions/types.html#types) there are no constrains on the range of the `number` type.

### Actual Behavior
```
> floor((pow(16, 32) / 4))
-9223372036854775808
> ceil((pow(16, 32) / 4))
-9223372036854775808
```

### Steps to Reproduce
Use `terraform console` to run the above examples.

### References
From my outsider perspective #26365 seems like it could be related?

**Comment by jbardin** [comment]
Thanks @wosc, this is a similar issue to #26365, and the upstream fix does note that there are other functions like `floor` and `ceil` which are still likely to have the same problem with values outside of the precision of double precision floats. We can open an issue upstream if there isn't one already to make sure there's a placeholder to check these.

**Comment by apparentlymart** [comment]
Interestingly, when I first implemented some operations like these in the upstream library I had it set up to return an error if given a value that was out of range for what the implementation could handle. I'm not sure if that error got weakened on the same functions in the meantime or if these were later additions that took a different approach. I _think_ these ones were ones that originated in the Terraform v0.11 tree as HIL functions and so perhaps we just ported them literally, where HIL itself only supported float64 precision.

Either way, it would be a breaking change to make these an error now and obviously it's better to make it work with the full precision if we can. Therefore I think the challenge now is to write arbitrary-precision implementations of these functions that are performant enough for configuration evaluation purposes (a relatively loose performance constraint), though perhaps in the meantime it would be wise to document that these particular functions (any that are left after the changes already made upstream) have a limited valid input range.
